src/controllers/settings.js

```diff
@@ -2,16 +2,191 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import {
+  getUserEmails,
+  resetUnverifiedEmailAddress,
+  addSubscriberUnverifiedEmailHash,
+  removeOneSecondaryEmail,
+  getEmailById,
+  verifyEmailHash
+} from '../db/tables/email_addresses.js'
+
+import { setAllEmailsToPrimary } from '../db/tables/subscribers.js'
+
+import { fluentError, getMessage } from '../utils/fluent.js'
+import { sendEmail, getVerificationUrl, getUnsubscribeUrl } from '../utils/email.js'
 import { mainLayout } from '../views/main.js'
 import { settings } from '../views/partials/settings.js'
+import AppConstants from '../app-constants.js'
+import { getBreachesForEmail } from '../utils/hibp.js'
+import { generateToken } from '../utils/csrf.js'
+
+import { getTemplate } from '../emails/templates/email-2022.js'
+import { verifyPartial } from '../emails/partials/email-verify.js'
+
+async function settingsPage (req, res) {
+  const emails = await getUserEmails(req.session.user.id)
+  // Add primary subscriber email to the list
+  emails.push({
+    email: req.session.user.primary_email,
+    sha1: req.session.user.primary_sha1,
+    primary: true,
+    verified: true
+  })
+
+  const breachCounts = new Map()
+  emails.forEach((email) => breachCounts.set(email.email, 0))
+
+  const allBreaches = req.app.locals.breaches
+  for (const email of emails) {
+    const breaches = await getBreachesForEmail(
+      email.sha1,
+      allBreaches,
+      true,
+      false
+    )
+    breachCounts.set(email.email, breaches.length)
+  }
 
-function settingsPage (req, res) {
   const data = {
-    fxaProfile: req.user.fxa_profile_json,
-    partial: settings
+    partial: settings,
+    emails,
+    breachCounts,
+    limit: AppConstants.MAX_NUM_ADDRESSES,
+    csrfToken: generateToken(res)
   }
 
   res.send(mainLayout(data))
 }
 
-export { settingsPage }
+async function addEmail (req, res) {
+  const sessionUser = await req.user
+  const email = req.body.email
+  // Use the same regex as HTML5 email input type
+  // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#basic_validation
+  const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/
+
+  if (!email || !email.match(emailRegex)) {
+    throw fluentError('user-add-invalid-email')
+  }
+
+  if (sessionUser.email_addresses.length >= AppConstants.MAX_NUM_ADDRESSES) {
+    throw fluentError('user-add-too-many-emails')
+  }
+
+  _checkForDuplicateEmail(sessionUser, email)
+
+  const unverifiedSubscriber = await addSubscriberUnverifiedEmailHash(
+    req.session.user,
+    email
+  )
+
+  await _sendVerificationEmail(unverifiedSubscriber.id)
+
+  // TODO: what should we return to the client?
+  return res.json('Resent the email')
```

Comment:
Resent? Judging by the name of the function `addEmail`, it's an universal add no?

As discussed in eng meeting, we want to get to a generic format, something like. This refactor work can certainly be in a different PR/jira task though!
```
{
  success: true,
  status: 200,
  message: '',
  stack (only dev/debug): {}
}
```

Reply:
Ah good catch, this is a holdover from the [V1 user controller](https://github.com/mozilla/blurts-server/blob/96fae648cd5428372e2401c4f964fa19ea857be6/controllers/user.js#L68-L69).

I'll go over the various responses and make them consistent for now, we discussed centralizing this and using schema validation etc. but I will file a follow-up for that.

Reply:
(note that the old `user` controller duplicated the function for sending and re-sending email verification, I consolidated that since I didn't see any important differences, but I ended up keeping the API response from the latter apparently)

Reply:
The error handler function (which should be triggered if we throw any uncaught exceptions) looks like it matches this format: https://github.com/mozilla/blurts-server/blob/8d941c5572bd07ae28121d457ff39dd3321b0c3f/src/middleware/error.js#L22-L32

So I think it's really the success cases we need to handle, I'll do it for now. We probably don't need `stack` for success cases I don't think.